doc/architectural/background-concepts.md

```diff
@@ -1,32 +1,824 @@
+we could not prove that the function never returns 0. In fact, this is
+correct: Calculating factorial(200) will yield 0, since the 64-bit
+integer fac overflows.
+
+Nevertheless, let us consider what would happen in a mathematically ideal
+world (e.g., if we used big integers). In that case, we would get
+"not 0" * "not 0" = "not 0", "not 0" + x = "not 0" and
+x + "not 0" = "not 0". Running the abstract interpretation with these
+semantics, we find that if we start BB3 with variable map N, we get
+variable map N at the end as well, so we end up with variable map N
+at the end of BB4 - but this means that fac maps to "not 0"!
+
+The algorithm sketched above is called a fixpoint algorithm because we
+keep propagating until we find that applying the transformers does not
+yield any new results (which, in a mathematically precise way, can be
+shown to be equivalent to calculating, for a specific function f, an x such
+that f(x) = x).
+
+This overview only describes the most basic way of performing abstract
+interpretation. For one, it only works on the procedure level (we speak of
+*intra-procedural analysis*); there are various ways of extending
+it to work across function boundaries, yielding *inter-procedural analysis*.
+Additionally, there are situations where it can be helpful make a variable map
+more abstract (widening) or use information that can be gained from
+various sources to make it more precise (narrowing); these advanced
+topics can be found in the literature as well.
+
+\section Glossary_section Glossary
+
+# Instrument
+
+To instrument a piece of code means to modify it by (usually) inserting new
+fragments of code that, when executed, tell us something useful about the code
+being instrumented.
+
+For instance, imagine you are given the following function:
+```C
+int aha (int a)
+{
+   if (a > 10)
+      return 1;
+   else
+      return -1;
+}
+```
+
+and you want to design an analysis that figures out which lines of code will
+be covered when `aha` is executed with the input `5`.
+
+We can *instrument* the code so that the function will look like:
+```C
+int aha (int a)
+{
+   __runtime_line_seen(0);
+   if (a > 10) {
+      __runtime_line_seen(1);
+      return 1;
+   } else {
+      __runtime_line_seen(2);
+      return -1;
+   }
+   __runtime_line_seen(3);
+}
+```
+
+All we have to do now is to implement the function
+`void __runtime_line_seen(int line)` so that when executed it logs somewhere
+what line is being visited. Finally we execute the instrumented version of
+`aha` and collect the desired information from the log.
+
+More generally speaking, and specially within the CPROVER code base,
+instrumenting the code often refers to modify its behavior in a manner that
+makes it easier for a given analysis to do its job, regardless of whether that
+is achieved by executing the instrumented code or by just analyzing it in some
+other way.
+
+# Flattening and Lowering
```

Comment:
Can we turn this (and probably others) into a doxygen `\subsection` so it can be referenced from else where